### PR TITLE
Implement Field

### PR DESCRIPTION
**Pull request by mikeywaites**
* Provide base Field class
* Provide library of built in Field types ``String`` ``Integer`` etc
* Provide pipeline system and pipeline iterator functionality

### CODE REVIEW

File: kim/utils.py [hunk 1]
```diff
@@ -18,3 +18,15 @@ def set_creation_order(instance):
...
+def attr_or_key(obj, name):
+    """attempt to use getattr to access an attribute of obj, if that fails
+    assume obj support key based look ups like a dict.
+
+    """
+
+    try:
+        return getattr(obj, name)
+    except AttributeError:
+        return obj.get(name)
```
Comment (jackqu7) [inline kim/utils.py]: add another except clause here to raise a helpful error telling people we don't know WTF they're doing?

File: kim/field.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+    def __init__(self, **field_opts):
+        """Construct a new instance of field.  Each field accepts a set of
+        kwargs that will be passed directly to the fields
+        defined ``opts_class``.
+        """
+
+        self.opts = self.opts_class(**field_opts)
+
+        set_creation_order(self)
```
Comment (jackqu7) [inline kim/field.py]: whats the idea here?

File: kim/field.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+class FieldOpts(object):
+
+    def __init__(self, **opts):
+
+        self._opts = opts
+
+        self.required = opts.pop('required', False)
+        self.default = opts.pop('default', None)
+        self.source = opts.pop('source', None)
+        self.name = opts.pop('name', self.source)
```
Comment (jackqu7) [inline kim/field.py]: Mapper metaclass should set attribute_name as name it was declared on the Mapper as, then:
```python
self.name = opts.pop('name', self.attribute_name)
self.source = opts.pop('source', self.name)
```

File: kim/field.py (hunk 2 above)
Comment (mikeywaites) [inline kim/field.py]: What bit do you mean specifically? Are you referring to `set_creation_order(self)`

**Comment by mikeywaites** [comment]
@jackqu7  can you take a look at 66a32e8 

We also need to agree on the process in which field.name/field.attribute name are set during Mapper instantiation.

**Comment by mikeywaites** [comment]
Ok so pretty much a carbon copy for what we were doing before.  If you can just sanity check the assumptions being made in 66a32e8 and 8f8465b @jackqu7.

File: kim/field.py (hunk 2 above)
Comment (jackqu7) [inline kim/field.py]: yeah i think you did explain it to me at the time though

**Pull request merged**